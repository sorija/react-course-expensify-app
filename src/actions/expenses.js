import uuid from 'uuid';
import database from '../firebase/firebase';

// Action generators

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  // can return a function because of redux-thunk
  return (dispatch) => {
    // destructure (same as destructuring in function's arg)
    const {
      description = '', 
      note = '', amount = 0, 
      createdAt = 0 
    } = expenseData;
    // save data
    const expense = { description, note, amount, createdAt };
    // we return a promise, so then we can chain it in the test case
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
