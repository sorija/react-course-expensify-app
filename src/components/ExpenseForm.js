import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//goal: use local component's state to track the changes to all the inputs
//only when form is submitted, we sent it up to redux and do something with the data
//BUT we DONT DISPATCH the action in this component-we want it to be reusable!
//so we just passed the data out of ExpenseForm (by calling a prop that gets passed in from the parent
export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      //.amount is a number in cents, we want a string with decimal places
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      //.moment() gets current time
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      description: props.expense ? props.expense.description : '',
      calendarFocused: false,
      error: '',
      note: props.expense ? props.expense.note : ''
    };
  };
  onDescriptionChange = (e) => {
    //e.target needs to be pulled first!
    //this.setState(() => ({ description: e.target.value })); would not work without .persist
    const description = e.target.value;
    this.setState(() => ({ description })); //description: description
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    //!amount allows user to clear the input
    //match a price up to 2 decimal points
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    //if statement prevents user from being able to clear the date input
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault(); //prevent full page refresh
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }));
    } else {
      //clear the error
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        //amount is a string, so we need to format it
        //parseFloat keeps decimal place and turns into real number base 10
        //multiply by 100 because we work with cents
        amount: parseFloat(this.state.amount, 10) * 100,
        //createdAt is a Moment stamp->have to format it to miliseconds
        //.valueOf is a method provided by Moment
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {/*conditionally render the error*/}
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input  
            type="text"
            placeholder="Description"
            autoFocus //automatically puts cursor/focus on the form
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text" //'text' instead of 'number' to enforece currency specific restrictions
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false } //all dates available
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
};