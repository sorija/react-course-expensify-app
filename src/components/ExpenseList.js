import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';


//map to iterate
//render new instance of ExpenseList component for each
//(this export is for testing)
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No expenses</span>
          </div>
        ) : (
          //map an array of objects and return instances of ExpenseListItem
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id}{...expense}/>
          })
        )
      }
    </div>
  </div>
);

// {
//   props.options.map((option, index)=> (
//     <Option
//       key={option}
//       optionText={option}
//       count={index + 1}
//       handleDeleteOption={props.handleDeleteOption}
//     />
//   ))
// }

const mapStateToProps = (state) => {
  return {
    //access filtered, sorted array
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);

//...is pro version of: 
// export default connect((state) => {
//   return {
//     expenses: state.expenses
//   }
// })(ExpenseList);

//..which is same as :
//const ConnectedExpense list = connect()...