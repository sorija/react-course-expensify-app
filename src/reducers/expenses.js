// Expenses Reducer

// variable for default state value
const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action ) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      // add expense to array without changing the state(original array)
      return [...state, action.expense]; //same as: return state.concat(action.expense)
    case 'REMOVE_EXPENSE':
      // remove item without changing original state
      // array 'expenses' gets destructurized to get id
      return state.filter(({ id }) => id !== action.id);
      // same as: return state.filter((expense) => expense.id !== action.id);
    case 'EDIT_EXPENSE':
      // iterate all expenses in the array
      return state.map((expense) => {
        // check if id of the current expense matches the one we try to find
        if (expense.id === action.id) {
          //if match
          return {
            ...expense, // grab all properties from existing object...
            ...action.updates //...override properties that were passed in
          };
        } else {
          return expense; // no change
        };
      });
    case 'SET_EXPENSES':
      return action.expenses;
    default:
      return state;
  }
};

export default expensesReducer;
