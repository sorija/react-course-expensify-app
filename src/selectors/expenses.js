import moment from 'moment';
import { create } from 'domain';
// Get visible expenses

// destructure 'filters' 
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  // determine if particular expense should be included 
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    //dont filter based on this value (return true) if no start/end date provided
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true ;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    // check if expenses.description has the text variable string inside of it
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // if all true then item will be kept in the 'expenses' array
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      // -1 means 'a' comes first, 1 means 'b' comes first
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy == 'amount') {
      return a.amount < b.amount ? 1 : -1; //greater amount comes first
    }
  });
};

export default getVisibleExpenses;
