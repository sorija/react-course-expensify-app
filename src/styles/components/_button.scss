.button {
  background: $bright-blue;
  border: none;
  color: white;
  display: inline-block;
  font-size: $font-size-large;
  font-weight: 300;
  line-height: 1;
  padding: $S-size;
  text-decoration: none;
}

.button--link {
  background: none;
}
