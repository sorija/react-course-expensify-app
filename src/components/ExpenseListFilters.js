import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter , sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  //it will get called by the 'react-dates' with the object that contains start/end date, so we can just destructure it
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    //need to update the redux store,
    //because the input's value is always gonna be the text value on the store
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div>
        {/*set value to whatever the current search text is*/}
        <input 
          type="text" 
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select 
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        {/*allow user to pick the range of dates they want to see expenses for*/}
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
    </div>
    );
  }
};
 
const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);