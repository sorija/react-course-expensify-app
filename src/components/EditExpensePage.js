import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

// Refactored code with class based component
// allows to pull out in-line functions to methods...
//... so we don't have to redefine them every time the component gets rendered

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    //'expense' are the updates
    // props.dispatch(editExpense(props.expense.id, expense));
    this.props.editExpense(this.props.expense.id, expense)
    this.props.history.push('/');
  };
  onRemove = () => {
    // props.dispatch(removeExpense({ id: props.expense.id }));
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button 
      onClick={this.onRemove}
      >
      Remove
      </button>
      </div>
    );
  }
};

//need to connect cuz we want to give the component the current expense object

//we are searching expenses array (that lives in 'state') for an expense that id matches props.match.params.id
//router renders the hoc, hoc passes the props throught and allows to add new ones
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  // define props that will be able to call dispatch
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

//=========================================//
//PREVIOUS CODE
// import React from 'react';
// import { connect } from 'react-redux';
// import ExpenseForm from './ExpenseForm';
// import { editExpense, removeExpense } from '../actions/expenses';

// const EditExpensePage = (props) => {
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push('/');
//         }}
//       />
//       <button 
//       onClick={() => {
//         props.dispatch(removeExpense({ id: props.expense.id }));
//         props.history.push('/');
//       }}
//     >
//     Remove
//     </button>
//     </div>
//   );
// };

// const mapStateToProps = (state, props) => {
//   return {
//     expense: state.expenses.find((expense) => expense.id === props.match.params.id)
//   };
// };

// export default connect(mapStateToProps)(EditExpensePage);
