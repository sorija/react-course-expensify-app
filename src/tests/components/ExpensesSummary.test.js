import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should correctly render ExpensesSummary with a single expense', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should correctly render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={15} expensesTotal={34092395}/>);
  expect(wrapper).toMatchSnapshot();
});
