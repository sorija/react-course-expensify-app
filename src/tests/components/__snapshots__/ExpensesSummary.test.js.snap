// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with a single expense 1`] = `
<div>
  <h2>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h2>
</div>
`;

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h2>
    Viewing 
    15
     
    expenses
     totalling 
    $340,923.95
  </h2>
</div>
`;
